components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp

```diff
@@ -13,18 +13,74 @@ namespace shoc {
 
 template<typename S, typename D>
 KOKKOS_FUNCTION
-void Functions<S,D>::diag_second_shoc_moments(const Int& shcol, const Int& nlev, const Int& nlevi, 
-                          const uview_1d<const Spack>& thetal, const uview_1d<const Spack>& qw, const uview_1d<const Spack>& u_wind, 
-                          const uview_1d<const Spack>& v_wind, const uview_1d<const Spack>& tke, const uview_1d<const Spack>& isotropy, 
-                          const uview_1d<const Spack>& tkh, const uview_1d<const Spack>& tk, const uview_1d<const Spack>& dz_zi, 
-                          const uview_1d<const Spack>& zt_grid, const uview_1d<const Spack>& zi_grid, const uview_1d<const Spack>& shoc_mix, 
-                          const uview_1d<const Spack>& wthl_sfc, const uview_1d<const Spack>& wqw_sfc, const uview_1d<const Spack>& uw_sfc, 
-                          const uview_1d<const Spack>& vw_sfc, const uview_1d<Spack>& thl_sec, const uview_1d<Spack>& qw_sec, 
-                          const uview_1d<Spack>& wthl_sec, const uview_1d<Spack>& wqw_sec, const uview_1d<Spack>& qwthl_sec, 
-                          const uview_1d<Spack>& uw_sec, const uview_1d<Spack>& vw_sec, const uview_1d<Spack>& wtke_sec, const uview_1d<Spack>& w_sec)
+void Functions<S,D>::diag_second_shoc_moments(const MemberType& team, const Int& nlev, const Int& nlevi, 
+       const uview_1d<const Spack>& thetal, const uview_1d<const Spack>& qw, const uview_1d<const Spack>& u_wind, 
+       const uview_1d<const Spack>& v_wind, const uview_1d<const Spack>& tke, const uview_1d<const Spack>& isotropy,
+       const uview_1d<const Spack>& tkh, const uview_1d<const Spack>& tk, const uview_1d<const Spack>& dz_zi, 
+       const uview_1d<const Spack>& zt_grid, const uview_1d<const Spack>& zi_grid, const uview_1d<const Spack>& shoc_mix, 
+       const Scalar& wthl_sfc, const Scalar& wqw_sfc, const Scalar& uw_sfc, const Scalar& vw_sfc, Scalar& ustar2, Scalar& wstar, 
+       const uview_1d<Spack>& isotropy_zi, const uview_1d<Spack>& tkh_zi, const uview_1d<Spack>& tk_zi, const uview_1d<Spack>& thl_sec, 
+       const uview_1d<Spack>& qw_sec, const uview_1d<Spack>& wthl_sec, const uview_1d<Spack>& wqw_sec, const uview_1d<Spack>& qwthl_sec, 
+       const uview_1d<Spack>& uw_sec, const uview_1d<Spack>& vw_sec, const uview_1d<Spack>& wtke_sec, const uview_1d<Spack>& w_sec)
 {
+  // This is the main routine to compute the second
+  // order moments in SHOC.
 
+  // Calculate surface properties needed for lower
+  //  boundary conditions
+  shoc_diag_second_moments_srf(wthl_sfc, uw_sfc, vw_sfc, ustar2, wstar);
+  team.team_barrier();
 
+  // Diagnose the second order moments flux, for the lower boundary
+  const auto nlevi_pack = ekat::npack<Spack>(nlevi)-1;
+  const int nlevi_indx  = (nlevi-1)%Spack::n;
+  Scalar wthl_s         = wthl_sec(nlevi_pack)[nlevi_indx];
```

Comment:
This code is significantly more complicated than it needs to be. If you just pass these outputs directly (they are taken by reference), then you don't need all these temporaries.

```
shoc_diag_second_moments_lbycond(wthl_sfc, wqw_sfc, uw_sfc, vw_sfc, ustar2, wstar,
          wthl_sec(nlevi_pack)[nlevi_indx], 
          wqw_sec(nlevi_pack)[nlevi_indx],
          uw_sec(nlevi_pack)[nlevi_indx],
          vw_sec(nlevi_pack)[nlevi_indx],
          wtke_sec(nlevi_pack)[nlevi_indx],
          thl_sec(nlevi_pack)[nlevi_indx],
          qw_sec(nlevi_pack)[nlevi_indx],
          qwthl_sec(nlevi_pack)[nlevi_indx]);
```